geode-lucene/src/main/java/org/apache/geode/cache/lucene/internal/distributed/LuceneQueryFunction.java

```diff
@@ -122,6 +128,98 @@ public void execute(FunctionContext<LuceneFunctionContext> context) {
         | PrimaryBucketException e) {
       logger.debug("Exception during lucene query function", e);
       throw new InternalFunctionInvocationTargetException(e);
+    } catch (LuceneIndexCreationInProgressException ex) {
+      handleException(ex, ctx);
+    }
+  }
+
+  private void handleException(LuceneIndexCreationInProgressException ex,
+      RegionFunctionContext ctx) {
+    PartitionedRegion userDataRegion = (PartitionedRegion) ctx.getDataSet();
+    // get local version
+    Version localVersion = ((InternalCache) ctx.getCache()).getDistributionManager()
+        .getDistributionManagerId().getVersionObject();
+    // get the remote members
+    Set<InternalDistributedMember> remoteMembers =
+        userDataRegion.getRegionAdvisor().adviseAllPRNodes();
+
+    if (!remoteMembers.isEmpty()) {
+      for (InternalDistributedMember remoteMember : remoteMembers) {
+        if (remoteMember.getVersionObject().ordinal() < localVersion.ordinal()) {
+          // reexecute but wait till indexing is complete
+          execute(ctx, true);
+          return;
+        }
+      }
+    }
+    // Will return the LuceneIndexCreationInProgressException as the new servers can handle this
+    // exception
+    throw ex;
+  }
+
+  public void execute(FunctionContext<LuceneFunctionContext> context, boolean waitOnRetry) {
```

Comment:
Would it make sense to have the regular execute call this method with the second parameter set to false?  The only difference is if the second parameter can be used to determine what type of exception handling should occur.